src/libzfs/py_zfs_vdev.c

```diff
@@ -1050,5 +1292,30 @@ py_zfs_vdev_t *init_zfs_vdev(py_zfs_pool_t *pool, nvlist_t *tree,
 	else
 		out->path = NULL;
 
+	if (is_spare && is_cache) {
+		PyErr_SetString(PyExc_RuntimeError,
+		    "A VDEV cannot be both spare and cache device");
+		Py_DECREF(out);
+		return (NULL);
+	}
+
+	if (is_spare)
+		out->class = PyUnicode_FromString(VDEV_TYPE_SPARE);
+	else if (is_cache)
+		out->class = PyUnicode_FromString(VDEV_TYPE_L2CACHE);
+	else {
+		if (strcmp(type, VDEV_TYPE_ROOT) == 0)
+			out->class = NULL;
+		else {
+			const char *bias = NULL;
+			if (nvlist_lookup_string(out->vdev_tree,
+			    ZPOOL_CONFIG_ALLOCATION_BIAS, &bias) == 0)
+				out->class = PyUnicode_FromString(bias);
+			else
+				out->class = PyUnicode_FromString("data");
+		}
+	}
+	Py_XINCREF(out->class);
```

Comment:
Is this leaking a reference? It looks like we're generating new unicode objects everywhere.

Reply:
Thanks, I have removed `Py_XINCREF(out->class)`.